packages/react-vapor/src/components/unsavedChangeModal/UnsavedChangeModalProvider.tsx

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+import {Button} from '../button/Button';
+import {ModalComposite} from '../modal/ModalComposite';
+
+const defaultModalTitle = 'Unsaved Changes';
+const defaultModalClasses = ['mod-prompt', 'mod-fade-in-scale'];
+const defaultModalDescription =
+    'Are you sure you want to leave this page without saving? All unsaved changes will be lost.';
+const defaultConfirmButtonText = 'Exit without applying changes';
+
+export interface IUnsavedChangesModalProviderProps {
+    isDirty: boolean;
+    modalTitle?: string;
+    className?: string[];
+    children: (props: {promptBefore: (callbackOnExit: () => any) => boolean}) => React.ReactNode;
+    unsavedChangesDescription?: string;
+    confirmButtonText?: string;
+}
+
+export const UnsavedChangesModalProvider: React.FunctionComponent<IUnsavedChangesModalProviderProps> = ({
+    isDirty,
+    children,
+    modalTitle = defaultModalTitle,
+    className = defaultModalClasses,
+    unsavedChangesDescription = defaultModalDescription,
+    confirmButtonText = defaultConfirmButtonText,
+}) => {
+    const [isOpen, setIsOpen] = React.useState(false);
+    const [resolve, setResolve] = React.useState(() => (shouldExitTheModal: boolean) => {
+        // do nothing
+    });
+
+    const promptBefore = (callback: () => any): boolean => {
+        if (isDirty) {
+            setIsOpen(true);
+
+            setResolve(() => (shouldExitTheModal: boolean) => {
+                setIsOpen(false);
+                shouldExitTheModal && callback();
+            });
+            return false;
+        }
+        return true;
+    };
+
+    return (
+        <>
+            {children({promptBefore})}
+            <ModalComposite
+                title={modalTitle}
+                classes={className}
+                modalHeaderClasses={['mod-confirmation']}
+                modalBodyClasses={['mod-header-padding', 'mod-form-top-bottom-padding']}
+                modalBodyChildren={<div>{unsavedChangesDescription}</div>}
+                modalFooterChildren={
+                    <>
+                        <Button name={confirmButtonText} onClick={() => resolve(true)} primary />
```

Comment:
Instead of `resolve(true/false)`, maybe we could have two functions, like `confirm()` and `cancel()`?

`cancel` is pretty much static, it should call `setIsOpen(false)`, then you could have `setOnConfirm(callback)` 🤔 

It would be much simpler

Reply:
:heavy_check_mark: :+1: